[Step 1] baem, 준호

Dylan-yoon opened the pull request:
@kiwi1023

# 😀코드
- func createThreeRandomInt() -> [Int]
  - 세개의 랜덤 숫자를 생성해 배열로 반환하는 함수
 
- determineStrikesBallsInTwoNumbers(_ computerNumbers: [Int], _ userNumbers: [Int]) -> (strike: Int, ball: Int)
  - 볼, 스트라이크 판정 하는 함수

- playGame()
  - 임의의 수를 생성 하여 판정 결과를 출력하는 함수

# 😱고민 및 조언을 얻고 싶은 부분
- 타입 선언을 명시적으로 하는 것이 좋은지, 혹은 암시적으로 하는것이 좋은지.
- 실행 예시처럼 출력하기위해 아래 코드와 같이 작성 했는데, 더 좋은 방법이 있는지.
``` swift 
userNumbers.map { String($0) }.joined(separator: " "))
```

Dylan-yoon committed:
docs: 순서도 작성

Dylan-yoon committed:
feat: createThreeRandomInt함수 생성

junho15 committed:
feat: compareResult함수 생성

Dylan-yoon committed:
feat: playGame 함수 구현

Dylan-yoon committed:
refactor: 컨벤션 수정 및 네이밍 변경

kiwi1023 commented:
안녕하세요 baem, 준호!
숫자게임 STEP1 고생많으셨습니다 ㅎㅎ
가능하시다면 커밋단위를 두분이 한번씩 번갈아가는 형태로 가시는걸 추천드려요~

😄리뷰전 당부의 말씀
---

이미 말씀 드린대로 제가 드린 리뷰는 정답이 아니니, 꼭 참고만 하시고 두분의 의견도 적극적으로 나눠주시길 바라겠습니다! 나아가 저희 모두에게 도움이 되는 프로젝트가 됬으면 좋겠습니다!

😃칭찬드리고 싶은점
---
- 함수네이밍 하실때 동사형 시작 아주 잘하셨습니다!
- 이건 개인적인 스타일이지만 저는 네이밍 할 경우 약어의 사용을 지양하는 편인데 여러분 모두 축약하지 않고 정확히 네임을 명시한것이 가독성을 올려주는거 같아 보기 좋습니다!!


🧐 질문에 대한 답변
---
- 타입 선언을 명시적으로 하는 것이 좋은지, 혹은 암시적으로 하는것이 좋은지.

두분이 처음 프로젝트를 시작하시기 전에 협의를 통하여 타입의 명확성을 위해 명시적으로 하는 것이 필요하다고 생각하셨다면, 명시적으로 하는 것이 좋을거고, 그 반대라고 생각을 하셨다면 암시적으로 하시는것도 또 하나의 방법이라고 생각합니다. 

하지만 앞으로 코딩을 하시다 보면 에러 및 혼란을 피하기 위해 꼭 명시를 해야하는 경우도 더러 발생할 수 있습니다. 이러한 경우에만 신경을 한번 더 써주시면 좋을 것 같습니다.

ex) 
1.선언과 초기화를 분리할 경우 
2.타입 추론으로 얻어지는 타입이 아닌, 다른 타입을 직접 지정할 필요가 있을 때

- 실행 예시처럼 출력하기위해 아래 코드와 같이 작성 했는데, 더 좋은 방법이 있는지.

고차함수의 활용 정말 잘하셨습니다!! 해당 코드의 큰 문제는 없어 보입니다. 제가 말씀드렸다시피 코드에 정답이 없기 때문에 두분께서 이 코드에 이유가 충분히 있다고 생각하시면 그대로 가셔도 될 것 같습니다.

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
 
 import Foundation
 
```
Foundation을 import 해주신 이유가 있으실가요?? Foundation이란 무엇이죠??

Dylan-yoon replied:
Foundation은 기본적으로 swift 언어를 사용함에 있어서
날짜, 시간,  문자열, 텍스트, 숫자, 정렬 등
기본 기능을 제공해주는 역할을 한다고 알고있습니다.!

junho15 replied:
네 Foundation은 iOS 앱의 운영체제 서비스와 기본 기능을 제공하는 프레임워크로 알고있습니다.

kiwi1023 replied:
그렇다면 그 기능이 여기서 필요할까요? 저게 import가 안되었다면 이 코드가 실행이 안될까요?🤔

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
 
+func createThreeRandomInt() -> [Int] {
+    var result: Set<Int> = Set<Int>()
+    while result.count < 3 {
```
Set을 활용한 중복숫자 제거 아주 좋습니다!

Dylan-yoon replied:
# 👍🏻

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
+}
+
+func playGame() {
+    var trialNumber: Int = 9
```
해당 변수를 playGame이라는 함수의 스코프 안에서 선언을 해주셨는데 이 변수는 해당 스코프 밖에서는 사용할 일이 없을까요?

Dylan-yoon replied:
처음에는 전역 변수로 사용했는데, 현재 저희 코드에서는 사용할 일이 없다고 판단하여 playGame 안에 넣어주었습니다.

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
+    return (strike: strike, ball: ball)
+}
+
+func playGame() {
```
해당 함수의 기능이 전체적으로 너무 크지 않나 싶어요! 조금더 분리를 할 수 있을거 같아요. 함수의 이름은 playGame 인데 어울리지 않는 기능들을 수행하고 있는거 같네요 ㅎㅎ

Dylan-yoon replied:
while 문 속의  if문 이하를 판단하는 함수를 분리하면 좋을 것 같다는생각이 듭니다. 
준호와 함께 변경해보도록 하겠습니다.

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
+    while true {
+        trialNumber -= 1
+        let userNumbers: [Int] = createThreeRandomInt()
+        print("임의의 수 : \(userNumbers.map { String($0) }.joined(separator: " "))")
```
해당 코드에 질문을 주셔서 굳이 한마디 더 드린다면  `userNumbers.map { String($0) }.joined(separator: " ")` 이 코드를 다른 변수에 할당한 다음 print를 해준다면 무엇을 print 하고 있는지 더욱 명확히 알 수 있을 것 같습니다!

Dylan-yoon replied:
네 확인했습니다.!

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
+}
+
+func determineStrikesBallsInTwoNumbers(_ computerNumbers: [Int], _ userNumbers: [Int]) -> (strike: Int, ball: Int) {
+    var ball: Int = 0
```
튜플을 함수에 활용 하신거 정말 좋습니다~!! 👍

Dylan-yoon replied:
처음에 제가 Array로 담아 두었는데, 준호가 변경해 주셨습니다.!👍🏻

kiwi1023 replied:
다양한 기능을 사용하신 것을 칭찬드린것이지 튜플이 정답이라는 뜻은 아니니 잘 고민하셔서 하시면 됩니다~!

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
-print("Hello, World!")
+playGame()
 
+func createThreeRandomInt() -> [Int] {
```
함수의 이름으로 타입명을 사용하지 않는 것을 추천드립니다~!

kiwi1023 replied:
https://swift.org/documentation/api-design-guidelines/

Dylan-yoon replied:
네 한번 더 읽어보고 숙지하도록 하겠습니다.

junho15 replied:
네 참고해서 함수명을 수정하겠습니다.

kiwi1023 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,57 @@
+    while true {
+        trialNumber -= 1
+        let userNumbers: [Int] = createThreeRandomInt()
+        print("임의의 수 : \(userNumbers.map { String($0) }.joined(separator: " "))")
```
map 과 compactMap의 차이점을 무엇일까요? 다른 고차함수는 무엇이 있을까요?

Dylan-yoon replied:
map은 배열의 각각의 element에 대한 변경을 할 때 사용합니다.
compactMap 은 map과 비슷하지만 Optional을 제외한 값을 반환합니다.
제가 사용했을 때는 ["1", "3", "사", "eight"] 를 Int로 변경하고자 할 때
map -> [1, 3, nil, nil] 
compactMap -> [1, 3] 
각각 반환됩니다

map, filter, reduce 는 사용을 해봤지만, compactMap, flatMap 은 아직 사용해본 적은 없습니다.

https://velog.io/@un1945/Swift-고차함수-compactMap-flatMap 
flatMap compactMap 다시찾아보던중, Kio 블로그도 자연스럽게 검색해서 찾아봤습니다~!

junho15 replied:
네 아직 모르는 유용한 고차함수가 많네요. 하나씩 찾아가며 공부해야겠습니다.

junho15 committed:
refactor: createThreeRandomInt 함수명 변경 외

createThreeRandomInt -> createThreeRandomNumbers
playGame 함수 기능 일부를 isWin함수를 만들어 분리함

kiwi1023 commented:
함수의 기능분화에 대해서 신경을 많이 써주셨으면 합니다~! 아주 중요한 문제 입니다! (왜중요할지도 고민해보시는 것도 좋을듯 합니다) 일단 여기서 머지 해드릴게요!
